### PR TITLE
[MM-29569] add https support and use config.json for service configuration

### PR DESCRIPTION
**Pull request by isacikgoz**
#### Summary
Add `https` support and use `config/config.json` for service configuration.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-29569

### CODE REVIEW

File: internal/server/server.go [hunk 1]
```diff
@@ -19,26 +20,49 @@ var (
...
+			tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
+			tls.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
+			tls.TLS_RSA_WITH_AES_128_GCM_SHA256,
+			tls.TLS_RSA_WITH_AES_256_GCM_SHA384,
 		},
 	}
 
+	s := &Server{
+		certFile: cfg.ServiceSettings.TLSCertFile,
+		keyFile:  cfg.ServiceSettings.TLSKeyFile,
```
Comment (agnivade) [inline internal/server/server.go:51]: I'd put the config itself in the `Server` and use the cert and key files from there. I also need the config for later stuff.

File: internal/server/config.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	"fmt"
+	"os"
+
+	"github.com/kelseyhightower/envconfig"
+)
+
+// Config is the configration for a bifrost server.
+type Config struct {
+	ServiceSettings *ServiceSettings  `split_words:"true"`
+	S3Settings      *AmazonS3Settings `split_words:"true"`
```
Comment (agnivade) [inline internal/server/config.go:17]: Let's remove the pointers. To avoid any possible races in config access.

File: internal/server/config.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+	Bucket          string
+	Region          string
+	Endpoint        string
+	Scheme          string
+}
+
+// ParseConfig reads the config file and returns a new *Config,
+// This method overrides values in the file if there is any environment
+// variables corresponding to a specific setting.
+func ParseConfig(path string) (*Config, error) {
```
Comment (agnivade) [inline internal/server/config.go:42]: Let's add a test for the config override stuff. And also return a struct value.

File: internal/server/server.go [hunk 4]
```diff
@@ -19,26 +20,49 @@ var (
...
+
+	server.TLSConfig = &tls.Config{
+		MinVersion: tls.VersionTLS12,
+		CipherSuites: []uint16{
+			tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
+			tls.TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,
+			tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
+			tls.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
+			tls.TLS_RSA_WITH_AES_128_GCM_SHA256,
+			tls.TLS_RSA_WITH_AES_256_GCM_SHA384,
```
Comment (agnivade) [inline internal/server/server.go:45]: We should either set what Filippo recommends [here](https://blog.cloudflare.com/exposing-go-on-the-internet/) or just go with the default. Messing around with ciphersuite order is not something I am confident enough to do. :)

Also, let's set `PreferServerCipherSuites: true`.

Comment (isacikgoz) [reply internal/server/server.go]: Ah OK, I was using `mattermost-server` as a basis 😅

File: internal/server/config_test.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+)
+
+func TestParseConfig(t *testing.T) {
+	dir, err := ioutil.TempDir("", "bifrost-test")
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		os.RemoveAll(dir)
+	})
+
+	t.Run("empty input should fail", func(t *testing.T) {
```
Comment (agnivade) [inline internal/server/config_test.go:22]: I'd also like to see a test with an environment variable overriding a config value.

File: internal/server/server.go [hunk 6]
```diff
@@ -19,26 +20,43 @@ var (
...
-			IdleTimeout:  30 * time.Second,
+func New(cfg Config) *Server {
+	server := &http.Server{
+		Addr:         net.JoinHostPort("0.0.0.0", cfg.ServiceSettings.Port),
+		ReadTimeout:  60 * time.Second,
+		WriteTimeout: 60 * time.Second,
+		IdleTimeout:  30 * time.Second,
+	}
+
+	server.TLSConfig = &tls.Config{
```
Comment (agnivade) [inline internal/server/server.go:36]: Hmm, why not declare this in the main struct declaration itself?

File: internal/server/server.go [hunk 7]
```diff
@@ -19,26 +20,43 @@ var (
 
 // Server contains all the necessary information to run Bifrost
 type Server struct {
+	cfg *Config
```
Comment (agnivade) [inline internal/server/server.go:23]: Let's also keep only a value here. Basically, I want the config to be initialized once and then read-only.

File: internal/server/config.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+
+	"github.com/kelseyhightower/envconfig"
+)
+
+// Config is the configration for a bifrost server.
```
Comment (agnivade) [inline internal/server/config.go:14]: configuration

**Review by agnivade — COMMENTED** [review]
Sorry, just a few more things to take care of :)

**Review by agnivade — APPROVED** [review]
Perfect :)

**Pull request merged**